dm/worker/server.go

```diff
@@ -137,20 +141,24 @@ func (s *Server) Start() error {
 		s.wg.Done()
 	}()
 
-	s.wg.Add(1)
-	go func() {
-		defer s.wg.Done()
-		// TODO: handle fatal error from observeSourceBound
-		//nolint:errcheck
-		s.observeSourceBound(s.ctx, revBound)
-	}()
+	s.startKeepAlive()
 
 	s.wg.Add(1)
 	go func() {
 		defer s.wg.Done()
-		// worker keepalive with master
-		// If worker loses connect from master, it would stop all task and try to connect master again.
-		s.KeepAlive()
+		for {
+			err1 := s.observeSourceBound(s.ctx, revBound)
+			if err1 == nil {
+				return
+			}
+			s.stopKeepAlive()
+			select {
+			case <-s.ctx.Done():
+				return
+			case <-time.After(time.Duration(s.cfg.KeepAliveTTL) * time.Second * 2):
```

Comment:
Can we start another keepalive only when we assure the last one is quitted? For example, using another waitgroup.